modules/calib3d/src/calibration.cpp

```diff
@@ -1498,26 +1496,20 @@ CV_IMPL double cvCalibrateCamera2( const CvMat* objectPoints,
 
             cvSub( &_mp, &_mi, &_mp );
 
-            if( _JtJ || _JtErr )
+            if( solver.state == CvLevMarq::CALC_J )
             {
-                cvGetSubRect( _JtJ, &_part, cvRect(0,0,NINTRINSIC,NINTRINSIC) );
-                cvGEMM( _Ji, _Ji, 1, &_part, 1, &_part, CV_GEMM_A_T );
-
-                cvGetSubRect( _JtJ, &_part, cvRect(NINTRINSIC+i*6,NINTRINSIC+i*6,6,6) );
-                cvGEMM( _Je, _Je, 1, 0, 0, &_part, CV_GEMM_A_T );
-
-                cvGetSubRect( _JtJ, &_part, cvRect(NINTRINSIC+i*6,0,6,NINTRINSIC) );
-                cvGEMM( _Ji, _Je, 1, 0, 0, &_part, CV_GEMM_A_T );
+                Mat JtJ(cvarrToMat(_JtJ)), JtErr(cvarrToMat(_JtErr));
 
-                cvGetRows( _JtErr, &_part, 0, NINTRINSIC );
-                cvGEMM( _Ji, _err, 1, &_part, 1, &_part, CV_GEMM_A_T );
+                // see HZ: (A6.14) for details on the structure of the Jacobian
+                JtJ(Rect(0, 0, NINTRINSIC, NINTRINSIC)) += _Ji.t() * _Ji;
+                JtJ(Rect(NINTRINSIC + i * 6, NINTRINSIC + i * 6, 6, 6)) = _Je.t() * _Je;
+                JtJ(Rect(NINTRINSIC + i * 6, 0, 6, NINTRINSIC)) = _Ji.t() * _Je;
 
-                cvGetRows( _JtErr, &_part, NINTRINSIC + i*6, NINTRINSIC + (i+1)*6 );
-                cvGEMM( _Je, _err, 1, 0, 0, &_part, CV_GEMM_A_T );
+                JtErr.rowRange(0, NINTRINSIC) += _Ji.t() * _err;
```

Comment:
ok, yet another question. I do not see where JtErr is cleared, to me it looks like it would be safe to clear JtErr (and probably JtJ too, because it's top-left 6x6 part is updated in the same way) before the loop over images.